### DISCUSSION TITLE
Heroku Hosting and Git Hub Pages

### DISCUSSION THREAD

**Discussion by pour-theoil** [nss-day-cohort-47/C47-Discussions#77] [Discuss a topic]
## Overview
A semi detailed discussion about how to host using GitHub Pages, as well as deploying a server using Heroku

_____
## Background
Over the last couple of weeks with the help of the instructor staff I have been able to host my Journal live. While it is far from build quality, learning how to build the application, all of the issues that come with using a remote server, and digging into Github Pages has been very rewarding. It seemed a little daunting from the beginning and I hope that describing the process will help to solidify the process. 

_____
## Heroku
If you are going to host the site globally, as well as fetch data, your local api server is not going to work for other viewers. There are loads of options for hosting; I choose Heroku because that's what Benda pointed be to and Bryan had experience deploying. 

1. Read the chapter on heroku <https://github.com/nss-day-cohort-47/client-side-mastery/blob/master/book-4-nashville-kennels/chapters/JSON_SERVER_HEROKU.md>
2. Sign up for **Heroku** <https://dashboard.heroku.com/>
2. Create a new app: 
3. Use terminal to create a new directory in your workspace and initialize it using the Heroku guide 
4. Add the data to be hosted and modify the package.json file
5. Push the code to git heroku master
6. Change the remoteurl at the top of the modules to the heroku server that you created
7. When you launch your app it takes up to 20 seconds for the data to load, so be  patient.

All of this is found in a supplemental chapter at the bottom of book 4

___
## GithubPages

I followed this guide <https://dev.to/yuribenjamin/how-to-deploy-react-app-in-github-pages-2a1f>

1. **Skip step 1** of the guide, you have likely already done this pushing to git
2. From the root of the project directory in terminal run `npm install gh-pages --save-dev`
3. Open the project in vscode so that you can modify specific scripts listed in step 3 of the guide. This creates a url for your home page as well as sets the scripts deploy your app remotely. Save the changes
4. **Skip step 4** of the guide, you have likely already done this pushing to git
5. In the root directory `npm run deploy` this will make a new branch on github for you to be able to use.
6. On github navigate to the directory that you are deploying. You should be able to look at the branches and see that a new gh-pages branch has been created!
7. In the repository navigate to settings, it will be in the nav bar possibly under the ...
8. Halfway down the page is the **GitHub Pages** header. In the source drop down select the gh-pages branch and save.

This "should" work, but it takes a while for everything to load... Your page will be at http://{username}.github.io/{repo-name} that you modified in step 3. Make sure that you fill in the correct information this website doesn't navigate from {username}.github.io

Good Luck!

**Comment by coltmay** [comment]
This is super cool, Logan.  I've been looking into [Coolify](https://coollabs.io/coolify/index.html), since I have an Ubuntu sever in my home anyway.  If I run into any issues, I may reach out, it may be similar to what you've done.

**Reply by pour-theoil** [reply]
Thank you. It seems like there are a million options, but super fun to have your own server.
